### DISCUSSION TITLE
Opt out of field auto-focus; error ordering

### DISCUSSION THREAD

**Discussion by Stanaland-Michael-bah** [edmundhung/conform#504] [Ideas]
We're planning a modernization of a very large and crusty React Router+Redux app, and I was looking at Conform as inspiration or a possible solution to dramatically improve the DX of our many forms.

We have some pretty strict accessibility requirements, and one of them is that form errors must be listed out at the top of the form in an alert (this WCAG pattern: https://www.w3.org/WAI/tutorials/forms/notifications/#listing-errors) ... this is in addition to the typical error message under the offending field. 

When the user submits a form with errors, the alert gets rendered and focus moves to the alert itself. But this conflicts with Conform's behavior of focusing the first invalid field. I spent a little time going through the docs and GitHub issues, and I don't see an easy way around that. 

---

The second challenge is around getting a list of invalid fields in the order they appear on the page.  Because `form.allErrors` is missing the input ids ad `form` and `fields` are proxy objects, it's not straight-forward to construct that list, but it is doable. A simplistic example (ignoring the fact that fields and have multiple errors) might look like this

```tsx
const errorFieldNames = Object.keys(form.allErrors);

const errors = errorFieldNames.map((name) => {
  return {
    name,
    id: fields[name].id,
    errors: fields[name].errors,
  };
});

return (
  <div role="alert" tabIndex={-1} ref={ref}>
    <h2>Correct these errors before proceeding</h2>
    <ul>
      {errors.map((error) => {
        return (
          <li key={error.id}>
            <a href={`#${error.id}`}>{error.errors[0]}</a>
          </li>
        );
      })}
    </ul>
  </div>
);
```

But we're relying on the order of `form.allErrors`, which doesn't really work currently. If `form.allErrors` was sorted and included the id, that would be amazing.

I know most other Conform users wont have this issue, and part of what makes Conform great is the simplicity and excellent accessibility behavior out of the box, but maybe something to think about for future releases.

**Comment by edmundhung** [comment]
Thanks for sharing! This does help me understand what people need if they would like to put all errors at the tops.

Is sorting the field names by their order on the dom tree enough？

**Reply by Stanaland-Michael-bah** [reply]
Yes, I would think so. I guess it would be possible for someone to do some flex-reverse thing that would make the DOM order not match the visual order on the screen, but generally that's an anti-pattern.

**Comment by lensbart** [comment]
Having the option to disable autofocus would be very welcome. In my case, I’m dealing with a form with input validation on blur. Some fields are mandatory and initially empty, so that if you blur by focusing another mandatory and empty field, you will get stuck in a perpetual focus/blur loop:

https://github.com/user-attachments/assets/8699d43a-fa12-475a-b9f7-fb9216d86917

Apart from this, I feel like it’s more user-friendly to not programmatically change the focus on blur, since it is unexpected and counters the user action:

https://github.com/user-attachments/assets/b21557f9-15ac-46d7-b53f-ad4e16dedf86

@edmundhung is this worth creating an issue for? Are you open to a quick PR that adds a boolean `focusOnError` to the form options?

Thanks.
